Fix Sync after sites refactor

ayumi opened the pull request:
See also: #10136 (sites refactor)

Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Test Plan:
0. Prepare a user profile with 1k+ bookmarks for Pyramid 0.
1. Start Pyramid 0 and enable Sync.
2. Start Pyramid 1 and join the Sync profile.
3. Wait for syncing to complete. To confirm, you can check both pyramids's consoles for "got 0 bookmarks".
4. Export bookmarks from both pyramids and compare with https://github.com/brave/browser-laptop/pull/10166 – they should be identical.

Reviewer Checklist:

Tests


- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

diracdeltas commented on app/browser/reducers/historyReducer.js:
```diff
@@ -60,19 +55,15 @@ const historyReducer = (state, action, immutableAction) => {
         if (Immutable.List.isList(action.get('historyKey'))) {
           action.get('historyKey', Immutable.List()).forEach((key) => {
             state = historyState.removeSite(state, key)
+            // TODO: Remove history site Sync
```
why is this a todo?

ayumi replied:
It's not a current Sync feature, so I think we should add it after the base Sync fixes are merged.

diracdeltas replied:
ah, it would be more clear if it were 'Sync history site removal', in this case i thought you were removing history site sync

NejcZdovc commented on app/browser/reducers/bookmarkFoldersReducer.js:
```diff
@@ -89,7 +78,9 @@ const bookmarkFoldersReducer = (state, action, immutableAction) => {
           break
         }
 
+        const folder = state.getIn(['bookmarkFolders', key])
```
Please use state helpers

NejcZdovc commented on app/browser/reducers/bookmarksReducer.js:
```diff
@@ -77,10 +68,8 @@ const bookmarksReducer = (state, action, immutableAction) => {
-          const destinationDetail = state.getIn(['sites', action.get('destinationKey')])
-          state = syncUtil.updateSiteCache(state, destinationDetail)
-        }
+        const destinationDetail = state.getIn(['bookmarks', action.get('destinationKey')])
```
Please use state helpers

NejcZdovc commented on app/browser/reducers/pinnedSitesReducer.js:
```diff
@@ -57,11 +55,8 @@ const pinnedSitesReducer = (state, action, immutableAction) => {
-          const newSite = state.getIn(['pinnedSites', action.siteKey])
-          state = syncUtil.updateSiteCache(state, newSite)
-        }
+        const newSite = state.getIn(['pinnedSites', action.siteKey])
```
Please use state helpers

NejcZdovc commented on app/sync.js:
```diff
@@ -284,18 +283,19 @@ module.exports.onSyncReady = (isFirstRun, e) => {
     if (typeof parentFolderId === 'number') {
       if (!folderToObjectId[parentFolderId]) {
-        const folderResult = bookmarkFolders.get(parentFolderId)
+        const folderResult = appState.getIn(['bookmarkFolders', parentFolderId])
```
Please use state helpers

NejcZdovc commented on js/state/syncUtil.js:
```diff
@@ -229,47 +228,94 @@ const applySyncRecord = (record) => {
 }
 
+const assignFolderIdsToRecords = (records, state) => {
+  let nextFolderId = bookmarkFoldersUtil.getNextFolderId(state.get('bookmarkFolders'))
```
Please use state helpers

NejcZdovc commented on js/state/syncUtil.js:
```diff
@@ -472,13 +523,13 @@ const findOrCreateFolderObjectId = (folderId, appState) => {
     appState = AppStore.getState()
   }
-  const folder = appState.getIn(['sites', folderId.toString()])
+  const folder = appState.getIn(['bookmarkFolders', folderId.toString()])
```
Please use state helpers

NejcZdovc commented on js/state/syncUtil.js:
```diff
@@ -472,13 +523,13 @@ const findOrCreateFolderObjectId = (folderId, appState) => {
     return objectId.toJS()
   } else {
-    return module.exports.newObjectId(['sites', folderId.toString()])
+    return module.exports.newObjectId(['bookmarkFolders', folderId.toString()])
```
Can we create constants file like `siteTags`. Because this way we are less error prone.

NejcZdovc commented on app/browser/reducers/bookmarkFoldersReducer.js:
```diff
@@ -75,10 +66,8 @@ const bookmarkFoldersReducer = (state, action, immutableAction) => {
-          const destinationDetail = state.getIn(['sites', action.get('destinationKey')])
-          state = syncUtil.updateSiteCache(state, destinationDetail)
-        }
+        const destinationDetail = state.getIn(['bookmarkFolders', action.get('destinationKey')])
```
This way we have state path defined only in one place `bookmarkFoldersState` and if we want to refactor anything it will help us a lot.

diracdeltas commented on js/state/syncUtil.js:
```diff
@@ -229,47 +228,94 @@ const applySyncRecord = (record) => {
+   * flush the buffer (buffered sites).
+   */
+  let buffer = []
+  let previousAppAction = null
```
safer to set this to a null function `() => {}`

ayumi commented:
@diracdeltas @NejcZdovc Thanks for your reviews. I made some updates! Please look again when you have a moment.

@NejcZdovc notably I added the `js/constants/stateConstants` file which looks like and I changed matching strings to use the new constants.
```js
const STATE_SITES = {
  BOOKMARKS: 'bookmarks',
  BOOKMARK_FOLDERS: 'bookmarkFolders',
  HISTORY_SITES: 'historySites',
  PINNED_SITES: 'pinnedSites'
}
```

NejcZdovc commented on app/browser/reducers/pinnedSitesReducer.js:
```diff
@@ -57,11 +56,8 @@ const pinnedSitesReducer = (state, action, immutableAction) => {
-          const newSite = state.getIn(['pinnedSites', action.siteKey])
-          state = syncUtil.updateSiteCache(state, newSite)
-        }
+        const newSite = pinnedSitesState.getSite(state, action.siteKey)
```
`action` is immutable, so it should be `action.get('siteKey')`. I see that I already made a mistake few lines above, will that in my PR :smiley:

NejcZdovc replied:
Fixed

ayumi commented:
@NejcZdovc have made the changes, thanks!

Sync is working mostly at this point; how do you feel about merging this branch into yours then I'll continue with Sync tests and other fixes?
